File: ai/runner/build_prompt.py
"""
Reads a .md prompt template from ai/agent/src/prompts/ and replaces
{{VARIABLE_NAME}} placeholders with values from a context dict.
"""

import re
from pathlib import Path

PROMPTS_DIR = Path(__file__).parent.parent / "agent" / "src" / "prompts"


def build_prompt(prompt_name: str, context: dict[str, str] = {}) -> str:
    path = PROMPTS_DIR / f"{prompt_name}.md"
    template = path.read_text(encoding="utf-8")

    for key, value in context.items():
        template = template.replace("{{" + key + "}}", value or "")

    return template


def list_prompt_variables(prompt_name: str) -> list[str]:
    path = PROMPTS_DIR / f"{prompt_name}.md"
    template = path.read_text(encoding="utf-8")
    return list(dict.fromkeys(re.findall(r"\{\{([A-Z_]+)\}\}", template)))

File: ai/agent/src/utils/buildPrompt.js
import { readFileSync } from "fs";
import { fileURLToPath } from "url";
import { dirname, join } from "path";

const __dirname = dirname(fileURLToPath(import.meta.url));
const PROMPTS_DIR = join(__dirname, "../prompts");

/**
 * Loads a prompt template from the prompts/ directory and replaces
 * {{VARIABLE_NAME}} placeholders with values from the provided context object.
 *
 * Missing variables in context are left as-is (so the caller can spot them).
 *
 * @param {string} promptName - filename without extension, e.g. "reproduce"
 * @param {Record<string, string>} context - key/value pairs to interpolate
 * @returns {string} the assembled prompt
 */
export function buildPrompt(promptName, context = {}) {
  const filePath = join(PROMPTS_DIR, `${promptName}.md`);
  let template = readFileSync(filePath, "utf-8");

  for (const [key, value] of Object.entries(context)) {
    const placeholder = new RegExp(`\\{\\{${key}\\}\\}`, "g");
    template = template.replace(placeholder, value ?? "");
  }

  return template;
}

/**
 * Lists the placeholder variables found in a prompt template.
 * Useful for validating that all required inputs are provided.
 *
 * @param {string} promptName - filename without extension
 * @returns {string[]} array of variable names (without the {{ }} brackets)
 */
export function listPromptVariables(promptName) {
  const filePath = join(PROMPTS_DIR, `${promptName}.md`);
  const template = readFileSync(filePath, "utf-8");
  const matches = [...template.matchAll(/\{\{([A-Z_]+)\}\}/g)];
  return [...new Set(matches.map((m) => m[1]))];
}
